src/main/java/cloud/eppo/ConfigurationRequestor.java

```diff
@@ -1,55 +1,148 @@
+                } else if (remoteFetchFuture != null && remoteFetchFuture.isDone()) {
+                  log.debug("Fetch beat the initial config; not clobbering");
+                } else {
+                  configurationStore.saveConfiguration(config);
+                  initialConfigSet = true;
+                }
+              }
+            });
+    return this.configurationFuture;
+  }
 
+  /** Loads configuration synchronously from the API server. */
+  void fetchAndSaveFromRemote() {
     log.debug("Fetching configuration");
-    byte[] flagConfigurationJsonBytes = requestBody("/api/flag-config/v1/config");
+
+    // Reuse the `lastConfig` as its bandits may be useful
+    Configuration lastConfig = configurationStore.getConfiguration();
+
+    byte[] flagConfigurationJsonBytes = client.get(FLAG_CONFIG_PATH);
     Configuration.Builder configBuilder =
         new Configuration.Builder(flagConfigurationJsonBytes, expectObfuscatedConfig)
             .banditParametersFromConfig(lastConfig);
 
-    if (configBuilder.requiresBanditModels()) {
-      byte[] banditParametersJsonBytes = requestBody("/api/flag-config/v1/bandits");
+    if (supportBandits && configBuilder.requiresBanditModels()) {
+      byte[] banditParametersJsonBytes = client.get(BANDIT_PARAMETER_PATH);
       configBuilder.banditParameters(banditParametersJsonBytes);
     }
 
-    configurationStore.setConfiguration(configBuilder.build());
+    configurationStore.saveConfiguration(configBuilder.build());
   }
 
-  private byte[] requestBody(String route) {
-    Response response = client.get(route);
-    if (!response.isSuccessful() || response.body() == null) {
-      throw new RuntimeException("Failed to fetch from " + route);
-    }
-    try {
-      return response.body().bytes();
-    } catch (IOException e) {
-      throw new RuntimeException(e);
+  /** Loads configuration asynchronously from the API server, off-thread. */
+  CompletableFuture<Void> fetchAndSaveFromRemoteAsync() {
+    log.debug("Fetching configuration from API server");
+    final Configuration lastConfig = configurationStore.getConfiguration();
+
+    if (remoteFetchFuture != null && !remoteFetchFuture.isDone()) {
+      log.debug("Remote fetch is active. Cancelling and restarting");
+      remoteFetchFuture.cancel(true);
+      remoteFetchFuture = null;
     }
+
+    remoteFetchFuture =
+        client
+            .getAsync(FLAG_CONFIG_PATH)
+            .thenApply(
+                flagConfigJsonBytes -> {
+                  synchronized (this) {
+                    Configuration.Builder configBuilder =
+                        new Configuration.Builder(flagConfigJsonBytes, expectObfuscatedConfig)
+                            .banditParametersFromConfig(
+                                lastConfig); // possibly reuse last bandit models loaded.
+
+                    if (supportBandits && configBuilder.requiresBanditModels()) {
```

Comment:
I wonder if we should rename `requiresBanditModels()` to `requiresUpdatedBanditModels()`, which (I think) is more explicit about what it is considering and makes the above code comment make more sense.